subsys/net/lib/sockets/socket_dispatcher.c

```diff
@@ -318,6 +328,26 @@ static int sock_dispatch_setsockopt_vmeth(void *obj, int level, int optname,
 			/* Native interface - use native socket implementation. */
 			fd = sock_dispatch_native(obj);
 		}
+	} else if ((level == SOL_TLS) && (optname == TLS_NATIVE)) {
+		const int *tls_native = (int *)optval;
+		struct dispatcher_context *ctx = (struct dispatcher_context *)obj;
+
+		if ((tls_native == NULL) || (optlen != sizeof(int))) {
+			errno = EINVAL;
+			return -1;
+		}
+
+		if (!is_tls(ctx->proto)) {
+			errno = ENOPROTOOPT;
+			return -1;
+		}
+
+		if (tls_native) {
```

Comment:
Should it be `if (*tls_native) {` maybe?

What about default, will it be native or non-native? What about passing value of 0, will there be a fallback from non-native to native during socket creation? Same for value of 1, will there be a fallback from native to non-native in case the former is "not supported" (seems that the current code does not assume any fallback, but just trying to confirm)?

Reply:
Fixed. The default depends on the socket priority. Typically, as offloaded sockets are processed first, it means the default will be offloaded socket (provided it accepts TLS socket types.

I wanted to keep the intermadiate "dispatcher" implementation as minimal as poosbile. I'm not sure what fallback is expected here? The option simply allows to enforce creation of a native TLS socket (we can't really use `SO_BINDTODEVICE`, as TLS socket, just as any other other native socket implementation is not bound to any interface). If not used, the dispatcher will rely on socket priorities and filtering function (just as it is right now). If we want to enforce an offloaded TLS creation, we can simply bind the socket to the offloaded interface.

Reply:
I was thinking about the case when TLS is not supported by offloaded socket implementation, but it (driver implementing offloaded sockets) still is requested to create TLS socket. Should such case be filtered out by calling `->is_supported()` callback, which at this point is not yet implemented correctly for existing offloaded socket drivers (they all return `true` without even checking if that is TLS or not)? Should #44656 fix that?

>I'm not sure what fallback is expected here?

I think I understand the mechanisms better now, so let's forget about that. Before, I was thinking about having a fallback in case some ciphersuite is not supported, but requested to be used, by offloaded socket implementation. That would require to collect lots of socket options (from setsockopt) and passing them to each network socket implementation. Let's stick with the simple approach.

One thing that still bothers me is whether native TLS should be default or offloaded TLS should be default. In theory there is more control over the native TLS stack, easier debug access (to the handshake phase), configuration of ciphersuites, etc. I would be weird if all applications relying on native TLS would now need to call `setsockopt()` with `TLS_NATIVE`. It is one step back from portability IMO. Could we introduce a Kconfig option to "force" native TLS? What do you think about that idea?

Reply:
> One thing that still bothers me is whether native TLS should be default or offloaded TLS should be default.

I also think the native TLS should be the default. Not sure if we need a kconfig option to force it if it is the default one. IMHO the user should do some coding (either adding some options in the code or setting kconfig options etc) in order to use the offload version.

Reply:
> Should https://github.com/zephyrproject-rtos/zephyr/pull/44656 fix that?

This is definitely a valuable addition to existing offloaded implmentations.

Reply:
The default choice is actually not really hardcoded in the networking stack with this proposal. It all depends on the socket priority that the offloaded implementation is registered with:
* if the offloaded socket have higher priority than the native one, and the offloaded implementation supports TLS, then this will be the default choice for the TLS sockets,
* if the offloaded socket have lower priority than the native one, the default will be native socket, no further action is needed by the application.

So the default choice is actually controlled by the offloaded engine provider. For example, for nRF91 and NCS examples we really prefer to default to the offloaded TLS socket, as use of the native TLS stack with nRF91 is actually an exception.

> I would be weird if all applications relying on native TLS would now need to call setsockopt() with TLS_NATIVE.

That's not really the case. `TLS_NATIVE` option is only needed if the offloaded engine supports TLS **and** has higher priority than the native one. I don't think any existing application would be impacted, as currently we don't even support "officially" native TLS socket with offloading (we only have workarounds in NCS, that were rejected upstream some time ago).

If the dispatcher is disabled, or the offloaded engine provider has decided for native sock to be the default by lowering priority, you don't need to use this option.

Reply:
>The default choice is actually not really hardcoded in the networking stack with this proposal. It all depends on the socket priority that the offloaded implementation is registered with

That is actually what I figured out yesterday, so thanks for confirming that!

>So the default choice is actually controlled by the offloaded engine provider. For example, for nRF91 and NCS examples we really prefer to default to the offloaded TLS socket, as use of the native TLS stack with nRF91 is actually an exception.

As you say, this is what **NCS examples** prefer. This is reasonable, as those should show the use case of offloaded TLS and its advantages.

However, for me as application / networking stack developer, I want to write as generic code as possible. This means that native TLS will be the default choice, because:
- it works the same on all platforms (compared to offloaded TLS that is platform/driver/networking chip specific),
- I will configure mbedTLS with supported ciphersuite and all other bits, as most platforms will still require it; I can do it once for all platforms, instead of bothering with each offloaded TLS one by one (configuration of offloaded TLS is usually limited, but it is still required to verify if ciphersuite is compatible and test if it actually works),
- native TLS allows to inspect TLS handshake and debug problems if some occur; debugging TLS is hard, but at least it is possible with open solution, such as mbedTLS (under Zephyr native TLS implementation)

IMHO open solution such as native TLS (with full software stack being open source) should be promoted by open projects such as Zephyr.

I have created related PR #44685, which allows to configure offloaded sockets priorities. So far I did not change priority value, so offloaded sockets (and offloaded TLS) is prioritized over native TLS sockets. In order to prioritize native TLS we can just swap priority values, which I think we should do (for the reasons I described above).

nRF91 examples can either increase priority of nRF91 modem offloaded sockets or introduce driver specific option that overwrites "Zephyr default" offloaded socket priority.

>If the dispatcher is disabled, or the offloaded engine provider has decided for native sock to be the default by lowering priority, you don't need to use this option.

Makes sense, thanks!

Reply:
Some generalization in terms of offloaded sockets priority makes sense I guess, thanks for the PR.